### PR TITLE
fix(bayes): Round expected improvement, bump sklearn version

### PR DESCRIPTION
**Pull request by hu-po**
Fixes [WB-8120](https://wandb.atlassian.net/browse/WB-8120?atlOrigin=eyJpIjoiNTE4M2ZkMmViMmMwNGZmNzg0YWQ5YmMxZjExYzE5NjUiLCJwIjoiaiJ9)

Rounds the expected improvement to 3 decimal places, which fixes an issue where the boundaries of parameter space were always being chosen in situations where every part of the parameter space should  have equal chance of being chosen due to an `np.argmax`.

Debugging journal: https://www.notion.so/wandbai/Debugging-scikit-learn-in-Sweeps-24258d1cacb648aa962782080d3d20ce

Slack thread: https://weightsandbiases.slack.com/archives/CS1GSNWLQ/p1658257216831819

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
 numpy>=1.15
 scipy>=1.5.4
 PyYAML
-scikit_learn==0.24.1
+scikit_learn>=0.24.2
```
Comment (jlzhao27) [inline requirements.txt:4]: should we just pin this to >= 1.x?

Comment (hu-po) [reply requirements.txt]: Actually, after changing it to 1.0 the py3.6 CI fails. It seems like the latest supported version for that is 0.24.2. @dannygoldstein wachu think?

Comment (jlzhao27) [reply requirements.txt]: oh womp:
```
Scikit-learn 0.20 was the last version to support Python 2.7 and Python 3.4. scikit-learn 0.23 and later require Python 3.6 or newer. scikit-learn 1.0 and later require Python 3.7 or newer.
```

I see, so we just have to unpin it and let it drift depending on the python version?

Comment (dannygoldstein) [reply requirements.txt]: the client supports 3.6 and since sweeps is an optional dependency of client (i.e., pip install wandb[sweeps]) we should continue supporting 3.6. we can do python version specific requirements, see e.g. https://pip.pypa.io/en/stable/reference/requirements-file-format/#example

Comment (dannygoldstein) [inline requirements.txt:4]: we should use python version dependent requirements here; 
```suggestion
scikit_learn>=0.24.2 ; python_version == 3.6
scikit_learn >= 1.0 ; python_version > 3.6
```

File: src/sweeps/bayes_search.py [hunk 2]
```diff
@@ -302,7 +299,18 @@ def next_sample(
...
+    # Check to see that the metric varies accross the samples
+    if np.all(np.isclose(filtered_y, filtered_y[0], atol=METRIC_VALUE_ATOL)):
+        logging.warning(
+            f"All instances of metric are within the minimum tolerance of {METRIC_VALUE_ATOL},"
+            + "the next sample will be a random sample within parameter space"
+        )
+        best_test_X_index = np.random.randint(0, test_X.shape[0] - 1)
+    else:
+        # Round for numerical stability
+        e_i = np.around(e_i, decimals=EXPECTED_IMPROVEMENT_SIGNIFICANT_DECIMAL_PLACES)
```
Comment (dannygoldstein) [inline src/sweeps/bayes_search.py:312]: with the above we should no longer need this. if we have to use this, we are relying on numerical instabilities 

```suggestion
```

File: src/sweeps/bayes_search.py [hunk 3]
```diff
@@ -302,7 +299,18 @@ def next_sample(
...
     else:
     """
-    best_test_X_index = np.argmax(e_i)
+
+    # Check to see that the metric varies accross the samples
+    if np.all(np.isclose(filtered_y, filtered_y[0], atol=METRIC_VALUE_ATOL)):
+        logging.warning(
+            f"All instances of metric are within the minimum tolerance of {METRIC_VALUE_ATOL},"
+            + "the next sample will be a random sample within parameter space"
+        )
```
Comment (dannygoldstein) [inline src/sweeps/bayes_search.py:308]: this will warn into the server logs which the user will never see. we should route warnings back to users

**Review by dannygoldstein — CHANGES_REQUESTED** [review]

**Comment by dannygoldstein** [comment]
we should also warn if users are running a sweep that has no valid versions of the metric

File: src/sweeps/bayes_search.py [hunk 4]
```diff
@@ -222,6 +215,7 @@ def next_sample(
         predicted_y: predicted value
         predicted_std: stddev of predicted value
         expected_improvement: expected improvement
+        warnings <List[str]>: warnings encountered
```
Comment (dannygoldstein) [inline src/sweeps/bayes_search.py:218]: I think `warnings` as you have it elsewhere in the code is just a string right now, should we update this or update warnings to be list of string?

File: src/sweeps/bayes_search.py [hunk 5]
```diff
@@ -249,6 +243,10 @@ def next_sample(
 
     filtered_X, filtered_y = filter_nans(sample_X, sample_y)
 
+    warnings = ""
+    if sample_y.shape[0] == 0:
```
Comment (dannygoldstein) [inline src/sweeps/bayes_search.py:247]: this isn't the right condition to filter on; by this point the y values have already been imputed with zero if they are missing a value of the metric. we should do this check in `_construct_gp_data`.

File: src/sweeps/bayes_search.py [hunk 6]
```diff
@@ -302,7 +303,14 @@ def next_sample(
         best_test_X_index = np.argmax(prob_of_improve)
     else:
     """
-    best_test_X_index = np.argmax(e_i)
+
+    # Check to see that the metric varies accross the samples
+    if np.all(np.isclose(filtered_y, filtered_y[0], atol=METRIC_VALUE_ATOL)):
```
Comment (dannygoldstein) [inline src/sweeps/bayes_search.py:308]: This is tricky, but here we want to check the gaussian process length scale and not that all y values are close to each other. With this we could get into a situation where we have only a few y values that were input by the user initially that are all close to each other, but the intervening parameter space has different values that vary on scales << 1. We should pull the kernel out of the fitted GP and check the length scale in each dimension, and just check that it is not above some threshold

File: src/sweeps/bayes_search.py [hunk 7]
```diff
@@ -444,11 +457,15 @@ def _construct_gp_data(
     if len(y) > 0:
         y[~np.isfinite(y)] = worst_metric
 
+    warnings = ""
+    if len(y) == 0:
+        warnings += "\nSweep has no valid samples of the metric."
```
Comment (dannygoldstein) [inline src/sweeps/bayes_search.py:462]: this condition does not capture all of the cases, see line 458 above. there can still be invalid values of the metric even if `len(y) > 0`. but by this point we have imputed them with `worst_metric`. we should move this block before line 457 but after line 456, and change it to 
```suggestion
    if y[~np.asarray(list(map(is_number, y)))].size == len(y):
        warnings += "\nSweep has no valid samples of the metric."
```

where `is_number` is defined in `run.py`. we may have to move that over from `run.py` to a utility file

**Review by dannygoldstein — COMMENTED** [review]
in the four failing cases can we please augment the unit tests to check that the warnings are appropriately routed through?

File: src/sweeps/bayes_search.py [hunk 8]
```diff
@@ -444,11 +457,18 @@ def _construct_gp_data(
     if len(y) > 0:
         y[~np.isfinite(y)] = worst_metric
 
+    warnings = ""
+    try:
+        if y[~np.asarray(list(map(is_number, y)))].size == len(y):
+            warnings += "\nSweep has no valid samples of the metric."
+    except TypeError:
+        warnings += "\nSweep has no valid samples of the metric."
```
Comment (dannygoldstein) [inline src/sweeps/bayes_search.py:465]:
```suggestion
    if len(y) == 0 or y[~np.asarray(list(map(is_number, y)))].size == len(y):
        warnings += "\nSweep has no valid samples of the metric."
```

**Review by dannygoldstein — APPROVED** [review]
